### PR TITLE
Dev/connection refactor

### LINKED ISSUE #533: use nox for running tests
we started testing using nox in https://github.com/ploomber/jupysql-plugin/pull/18/

we had some issues in the past with it (some weird errors when creating conda envs), but it has many benefits. For example, it'd make it easier to run tests with sqlalchemy 1.x and 2.x locally. So we'll see how it goes and decide if we implement it here.

### LINKED ISSUE #730: integration tests getting stuck
the problem is the same for all drivers, if the driver package is missing, the db hangs up

### LINKED ISSUE #731: replace `--live/--local` flags in integration tests
to run certain integration tests, we can pass the `--live/--local` flags. However, to determine if a test must run under `--live` or `--local`, we need to add a fixture to each test (e.g., to skip it during live mode). This is problematic because someone adding integration tests might not be aware of such flag.

As a consequence, we now have tests that are duplicated (they run in both the live and local CI job). I think a better alternative is to eliminate the flags and use pytest feature `-k` to selectively run tests.

As far as I know, the purpose of `--live` is to run snowflake tests, so we could do something like this:

```
pytest "src/tests/integration" -k "snowflake" 
```

this will only execute snowflake tests (allowing us to replace the `--live` flag.

Then, to run "all tests, except snowflake"

```
pytest "src/tests/integration" -k "not snowflake" 
```

### PR DESCRIPTION
**Pull request by edublancas**
## Describe your changes

The developer experience for JupySQL has been degrading over time: it became more difficult to run tests locally, to understand how db connections work, etc. I made a big refactor (and I'm planning to open 2-3 more PRs with the same objective, but this will be the one with the most file changes), here's the summary:

- broke the connection.py file into multiple ones
- standardized the sqlalchemy connections and dbapi ones to have a common, unified interface
- broke the run.py into multiple files
- added missing unit and integration tests, also added some that aren't passing yet since I discovered some bugs
- migrated the testing suite to use nox, in the next PR I'll simplify this so it's easier to run tests locally (closes #731, closes #533)
- renamed some attributes, methods and classes to have more descriptive names
- fixes some bugs and added tests
- fixed a bug that caused integration tests to get stuck (closes #730)

## Issue number

Closes #X

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--755.org.readthedocs.build/en/755/

### CODE REVIEW

**Comment by edublancas** [comment]
@tonykploomber @yafimvo @neelasha23 please prioritize reviewing this, as there are many changes, and other open PRs are likely to start encountering merge conflicts.

File: doc/community/developer-guide.md [hunk 1]
```diff
@@ -110,7 +110,7 @@ In our codebase, we manage connections to databases with a `Connection` object,
 
 ### Working with connections
 
-`Connection` should be exclusively used to manage database connections on the user's behalf and to obtain the current SQLAlchemy connection. We can access the current SQLAlchemy connection using `current.session`.
+`ConnectionManager` should be exclusively used to manage database connections on the user's behalf and to obtain the current connection. We can access the current connection using `current`.
```
Comment (neelasha23) [inline doc/community/developer-guide.md:113]: There's a line above this : `In our codebase, we manage connections to databases with a Connection object,` .. should this also be `ConnectionManager` ?

File: src/sql/_testing.py [hunk 2]
```diff
@@ -198,6 +198,8 @@ def database_ready(
             eng.close()
             print(f"{database} is initialized successfully")
             return True
+        except ModuleNotFoundError:
+            raise
```
Comment (neelasha23) [inline src/sql/_testing.py:202]: Do we want to perform any additional operations here?

File: src/tests/integration/test_connection.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                "server_version_info": ANY,
+            },
+        ],
+    ],
+    ids=[
+        "postgresql",
+        "duckdb",
+        "duckdb_native",
+    ],
+)
```
Comment (yafimvo) [inline src/tests/integration/test_connection.py:132]: Should we test more connections here? (mariadb, sqlite, snowflake, etc...)

File: src/tests/integration/test_duckDB.py [hunk 4]
```diff
@@ -1,64 +1,16 @@
...
-        ip.run_cell("%config SqlMagic.autocommit=False")
-        ip.run_cell("%sql CREATE TABLE weather (city VARCHAR,);")
-    # Check there is no message gets printed
-    assert caplog.record_tuples == []
-    # Check the tables is created
-    tables_out = ip.run_cell("%sql SHOW TABLES;").result
-    assert any("weather" == table[0] for table in tables_out)
-
-
 def test_dbapi_connection_sets_right_dialect(ip_with_duckDB_native):
```
Comment (yafimvo) [inline src/tests/integration/test_duckDB.py:11]: Can we add 1 more case when it's not duckdb? (I guess dialect can be `None` or `duckdb`)

File: src/tests/integration/test_generic_db_operations.py [hunk 5]
```diff
@@ -832,61 +833,286 @@ def test_sql_error_suggests_using_cte(ip_with_dynamic_db, request):
...
+            CREATE_TEMP_TABLE,
+            marks=pytest.mark.xfail(
+                reason="We're executing operations in different cursors"
+            ),
+        ),
+        pytest.param(
+            "ip_with_MSSQL",
+            CREATE_TABLE,
+            marks=pytest.mark.xfail(
+                reason="We need to close all existing result sets for this to work"
```
Comment (yafimvo) [inline src/tests/integration/test_generic_db_operations.py:993]: Should we create specific tests for mssql and oracale?

File: src/tests/integration/test_mssql.py [hunk 6]
```diff
@@ -1,4 +1,3 @@
-import pyodbc
```
Comment (yafimvo) [inline src/tests/integration/test_mssql.py:1]: Why was this removed from here and placed into the function?

File: src/tests/integration/test_run.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    with pytest.raises(psycopg2.errors.UndefinedTable):
+        with conn_two.cursor() as c:
+            c.execute(f"SELECT * FROM {name}")
+
+
+# TODO: duckdb-engine does not support isolation level so we need to test with
+# a database that does (but first verify that this is the case)
+
+
+@pytest.mark.xfail(reason="this is failing with sqlalchemy 1.x")
```
Comment (yafimvo) [inline src/tests/integration/test_run.py:94]: can we skip only if sqlalchemy<2 ?

File: src/sql/connection/connection.py [hunk 8]
```diff
@@ -484,95 +257,163 @@ def close_connection_with_descriptor(cls, descriptor):
...
+                        suggestion_str,
+                        PLOOMBER_DOCS_LINK_STR,
+                    ]
+                )
+            ) from e
+        except Exception as e:
+            raise cls._error_invalid_connection_info(e, connect_str) from e
 
-        cls.connections = {}
+        connection = SQLAlchemyConnection(engine, alias=alias)
```
Comment (yafimvo) [inline src/sql/connection/connection.py:310]: What happens if we pass a non-SQL alchemy connection string here? i.e. questdb.

**Review by yafimvo — CHANGES_REQUESTED** [review]
Looks good and much clearer! tested it locally and it works.

Added some comments

File: src/tests/test_run.py [hunk 9]
```diff
@@ -99,85 +87,82 @@ def test_handle_postgres_special(mock_conns):
...
-    monkeypatch.setattr("sql.run.ResultSet", mock_resultset)
-
-    output = run(mock_conns, "\\", config_pandas)
-    assert isinstance(output, type(mock_resultset.DataFrame()))
+# @pytest.mark.parametrize(
+#     "n, message",
+#     [
+#         [1, "1 rows affected.\n"],
+#         [0, ""],
+#     ],
```
Comment (neelasha23) [inline src/tests/test_run.py:116]: Commented code can be removed

File: src/tests/test_run.py [hunk 10]
```diff
@@ -99,85 +87,82 @@ def test_handle_postgres_special(mock_conns):
...
 
 @pytest.mark.parametrize(
-    "n, message",
+    "connection",
     [
-        [1, "1 rows affected.\n"],
-        [0, ""],
+        SQLAlchemyConnection(create_engine("duckdb://")),
+        SQLAlchemyConnection(create_engine("sqlite://")),
+        DBAPIConnection(duckdb.connect()),
```
Comment (neelasha23) [inline src/tests/test_run.py:129]: A bit confused here. Why are we testing DBAPIConnection with duckdb and sqlite when both support sqlalchemy?

File: src/sql/_testing.py (hunk 2 above)
Comment (edublancas) [reply src/sql/_testing.py:202]: no, we just need to raise the original exception. this allows the tests to fail quickly if they have missing dependencies

File: src/sql/connection/connection.py [hunk 11]
```diff
@@ -615,124 +456,266 @@ def get_curr_identifiers(self) -> list:
...
-        self.alias = alias
-        Connection.current = self
+        # calling init from AbstractConnection must be the last thing we do as it
+        # register the connection
+        super().__init__(alias=alias or self._connection_class_name)
 
-        # TODO: create an abstract class
-        self._result_sets = []
+        # TODO: delete this
+        self.name = self._connection_class_name
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/connection/connection.py:580]: I tried but then some tests started failing, because `.name` is used in other places. I'm gonna tackle this in the next PR.

File: src/sql/connection/connection.py [hunk 12]
```diff
@@ -615,124 +456,266 @@ def get_curr_identifiers(self) -> list:
...
+            "server_version_info": None,
+        }
+
+    @property
+    def url(self):
+        """Returns None since DBAPI connections don't have a url"""
+        return None
+
+    @property
+    def connection_sqlalchemy(self):
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/connection/connection.py:604]: no, we need this because both SQLAlchemyConnection and DBAPIConnection have a connection attribute but only the format has a connection_sqlalchemy.

So if any feature depends on sqlalchemy, you can do: `connection.connection_sqlalchemy` and it'll work if the user has a sqlalchemyconenction but it'll throw an error "this feature is only supported for sqlalchemy connections" if using a DBAPIConnection, this prevents the need to add conditionals everywhere to check whether you have a sqlalchemy connection or not

File: src/sql/connection/connection.py [hunk 13]
```diff
@@ -615,124 +456,266 @@ def get_curr_identifiers(self) -> list:
...
+        connection object
+        """
+        raise NotImplementedError(
+            "This feature is only available for SQLAlchemy connections"
+        )
+
+    # TODO: delete this, execution must be done via .execute
+    @property
+    def connection(self):
+        return self._connection
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/connection/connection.py:616]: we're still accessing the connection directly in a few places, I need to evaluate if we can delete it or no. will do this in the next PR

File: src/tests/integration/test_generic_db_operations.py (hunk 5 above)
Comment (edublancas) [reply src/tests/integration/test_generic_db_operations.py:993]: we can keep the tests for mssql and oracle here. to fix them I need to make a few changes, I'll work on them in an upcoming PR, for now I wanted to add the tests so we know we have to work on fixing this

File: src/tests/integration/test_mssql.py (hunk 6 above)
Comment (edublancas) [reply src/tests/integration/test_mssql.py:1]: because this allows us to run a subset of the integration tests. e.g., if I'm only interested in running postgres tests, then I shouldn't need to install pyodbc, since that's only required for mssql tests. it removes the complexity of requiring all dependencies, some of this are problematic to install

File: src/sql/connection/connection.py (hunk 8 above)
Comment (edublancas) [reply src/sql/connection/connection.py:310]: it'll fail. if you use `from_connect_str` and pass an invalid sqlalchemy string, it'll throw an error because connection strings assume you're using sqlalchemy connections

File: src/tests/test_run.py (hunk 10 above)
Comment (edublancas) [reply src/tests/test_run.py:129]: I realized that duckdb+sqlalchemy was generating a lot of issues, so we'll keep supporting it for retro compatibility but we'll now recommend users to use native connections with duckdb

File: src/tests/integration/test_connection.py (hunk 3 above)
Comment (edublancas) [reply src/tests/integration/test_connection.py:184]: [fixed](https://github.com/ploomber/jupysql/pull/755/commits/bd5151872d824f4da1848165973b4d34bb72fd69)

File: src/tests/integration/test_duckDB.py (hunk 4 above)
Comment (edublancas) [reply src/tests/integration/test_duckDB.py]: [fixed](https://github.com/ploomber/jupysql/pull/755/commits/493d8d933f67a319367a71397932822f033cbebe)

File: src/tests/integration/test_run.py (hunk 7 above)
Comment (edublancas) [reply src/tests/integration/test_run.py:99]: [fixed](https://github.com/ploomber/jupysql/pull/755/commits/a4a8569f22b9d8139e6337192288c644b90f9587)

File: src/tests/test_run.py (hunk 9 above)
Comment (edublancas) [reply src/tests/test_run.py]: [fixed](https://github.com/ploomber/jupysql/pull/755/commits/bbc5cd254e0d9e7579dfd3d92669207d1b29f4be)

File: src/sql/_testing.py (hunk 2 above)
Comment (neelasha23) [reply src/sql/_testing.py:202]: got it!

**Comment by edublancas** [comment]
please review again @neelasha23 @yafimvo

File: src/sql/run/pgspecial.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
+try:
+    from pgspecial.main import PGSpecial
+except ModuleNotFoundError:
+    PGSpecial = None
+
+from sql import exceptions
+
+
+def handle_postgres_special(conn, statement):
+    """Execute a PostgreSQL special statement using PGSpecial module."""
```
Comment (tonykploomber) [inline src/sql/run/pgspecial.py:10]: Do you think we can import PGSpecial only if it's called?

```suggestion
    """Execute a PostgreSQL special statement using PGSpecial module."""
    try:
    from pgspecial.main import PGSpecial
except ModuleNotFoundError:
    PGSpecial = None
```

File: src/tests/integration/test_run.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
```
Comment (tonykploomber) [inline src/tests/integration/test_run.py:99]:
```suggestion
@pytest.mark.skipif(
    SQLALCHEMY_VERSION == 1, reason="this is failing with sqlalchemy 1.x"
)
```

File: src/sql/run/run.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+        # TODO: checking for with isn't the best idea because it doesn't guarantee
+        # that the final one is a select statement
+        first_word_statement = statement.strip().split()[0].lower()
+
+        # in duckdb db "from TABLE_NAME" is valid
+        is_select = first_word_statement in {"select", "with", "from"}
+
+        if first_word == "begin":
+            raise exceptions.RuntimeError("JupySQL does not support transactions")
```
Comment (tonykploomber) [inline src/sql/run/run.py:50]: Just realize we don't have a test case to cover this condition, we can cover it in next PR

**Review by tonykploomber — COMMENTED** [review]
Small comment

File: src/sql/run/pgspecial.py (hunk 14 above)
Comment (edublancas) [reply src/sql/run/pgspecial.py:10]: I'm confused. this try ... except .. is already in line 1-4. applying this won't have any effect

File: src/sql/run/run.py (hunk 16 above)
Comment (edublancas) [reply src/sql/run/run.py:50]: we have a test. test_run.py, test named: test_sql_starts_with_begin

**Comment by edublancas** [comment]
Hey @neelasha23 @yafimvo please review again. We want to merge this asap so we don't delay other PRs

**Comment by neelasha23** [comment]
> Hey @neelasha23 @yafimvo please review again. We want to merge this asap so we don't delay other PRs

Had added one [comment](https://github.com/ploomber/jupysql/pull/755#discussion_r1273828747) @edublancas 
Rest looks ok

**Review by yafimvo — APPROVED** [review]
looks good from my end. added 2 small comments.

File: src/tests/test_run.py (hunk 10 above)
Comment (edublancas) [reply src/tests/test_run.py:116]: it isn't really that DBAPIConnection optimizes connections, it depends on the driver.
